### ISSUE TITLE
Potential inconsistent behaviour with apiserver apiextensions (custom resources) and 404s

### ISSUE THREAD

**Issue by kapad** [kubernetes/kubernetes#115241]
### What happened?

I'm working on a system the extends the kubernetes api server. 
This system consists of a kubernetes cluster, with several custom resource definitions and custom resources of each kind. 
These resources are consumed by _in-cluster_ and _external clients_. The difference between the two types of clients is that the in-cluster client uses a cached client for reads, while external clients use the REST API for reads as well as writes. 

Including the kubernetes components and other components that are deployed, the system has a few dozen in-cluster clients (including kubelets). The number of external clients is > 5000. 

Recently, there was an issue with all the k8s-apiservers of the cluster restarting at the same time. The in-cluster clients failed during this time window, eventually resulting in container restarts. And this was expected. 

The **potential inconsistent behaviour is in relation to the external clients**. Per system logs _(apiserver audit logs)_, around the time of the incident, the following two behaviours were observed
* external clients receiving a 5xx response, (expected)
* external clients receiving 404s for resources that exist. 

Most of the external clients got 5xx's, but there were two instances where the response was a 404. Even in the case of these two instances, the clients received 200 responses within ~30 seconds of receiving a 404.

During the incident window, the restarts of the k8s apiservers were observed at different times over a ~30s window. No delay was observed in the container restarts, and each apiserver pod was back up and receiving traffic within 30-40 seconds (or sooner). 
* per the audit logs, the same external client received, 200, 404, 200, 404, 200, 200, 200, .... (at ~30s intervals). 
* the audit logs are missing any information regarding the apiserver instance that emitted the log. 
* it's likely that the different 404s were seen by the same external client (for the same resource), for responses from two different apiserver instances

### What did you expect to happen?

All external clients should've received 5xx response codes while the apiservers were down.

### How can we reproduce it (as minimally and precisely as possible)?

* create k8s cluster. 
* add several custom resource definitions, and several resources of each kind
* (may or may not be relevant: etcd size is approximately 4 GB) 
* run a few clients that use the k8s REST API to query for different resources that were created earlier
* (overall rate of query ^ should be high, so that the intermittent inconsistency is observed). 
* shut down all api servers.
* (clients continue: start seeing 5xx)
* restart the api servers. 
* watch for 404s in client logs
* since clients are only requesting resources that are already created, 404 is not an expected response. 

### Kubernetes version

<details>

_ignore client version, that's from my local system_

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:33:49Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.10", GitCommit:"a7a32748b5c60445c4c7ee904caf01b91f2dbb71", GitTreeState:"clean", BuildDate:"2022-02-16T11:18:16Z", GoVersion:"go1.16.14", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

The system described in this issue is a kubernetes cluster deployed on Azure compute nodes. 

</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ uname -a
Linux cp-vmss000001 5.15.0-1019-azure #24~20.04.1-Ubuntu SMP Tue Aug 23 15:52:52 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
/sig api-machinery

> run a few clients that use the k8s REST API to query for different resources that were created earlie

These resources are CustomResources?

/cc @liggitt @deads2k

**Comment by liggitt** [comment]
> external clients receiving 404s for resources that exist.

were the 404s Kubernetes responses (contains v1 Status body content with a Kubernetes-constructed message) or just raw 404 responses?

> * per the audit logs, the same external client received, 200, 404, 200, 404, 200, 200, 200, .... (at ~30s intervals).

were these kubernetes API server audit logs?

> * the audit logs are missing any information regarding the apiserver instance that emitted the log.

how are the audit logs from the different servers aggregated? are the regular apiserver logs available? (those can log requests and response status codes)

**Comment by kapad** [comment]
> /sig api-machinery
> 
> > run a few clients that use the k8s REST API to query for different resources that were created earlie
> 
> These resources are CustomResources?
> 
> /cc @liggitt @deads2k

Yes. These resources are custom resources. Additionally, we see the inconsistent 404s only with custom resources. With core resources and core extensions, we see 5xx (expected during downtime), followed by 2xx (expected after apiservers are operational again).

**Comment by kapad** [comment]
> > external clients receiving 404s for resources that exist.
> 
> were the 404s Kubernetes responses (contains v1 Status body content with a Kubernetes-constructed message) or just raw 404 responses?
> 

At present, this information hasn't been collected. 
The client uses the controller-runtime client, and identifies 404s using the [`IgnoreNotFound`](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.14.1/pkg/client/interfaces.go#L200-L207) function. Would that mean that the responses contains the v1 Status body content with k8s api server constructed message? 

> > * per the audit logs, the same external client received, 200, 404, 200, 404, 200, 200, 200, .... (at ~30s intervals).
> 
> were these kubernetes API server audit logs?

Yes. These are kubernetes API server audit logs. 

> 
> > * the audit logs are missing any information regarding the apiserver instance that emitted the log.
> 
> how are the audit logs from the different servers aggregated? are the regular apiserver logs available? (those can log requests and response status codes)

I will need to check further on how the audit logs are aggregated. 

Yes. regular api server logs are available as well, but I wasn't able to find any references to the resources for which inconsistent 404s were observed in the apiserver logs.

**Comment by liggitt** [comment]
> The client uses the controller-runtime client, and identifies 404s using the [`IgnoreNotFound`](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.14.1/pkg/client/interfaces.go#L200-L207) function. Would that mean that the responses contains the v1 Status body content with k8s api server constructed message?

Not necessarily, that would also return true for generic 404 responses issued by something other than the Kubernetes API server, like a load balancer.

Does this client make discovery API calls as part of the call returning a NotFound error, and if those discovery calls return 404, would that surface as a 404?

> Yes. Regular api server logs are available as well. I wasn't able to find any references to the resources for which inconsistent 404s were observed in the apiserver logs though.

You don't see any 404 responses recorded in the apiserver log for those resources (either the individual objects or the discovery APIs for that API group or API group/version)?

**Comment by kapad** [comment]
> Not necessarily, that would also return true for generic 404 responses issued by something other than the Kubernetes API server, like a load balancer.

Got it. Thanks for confirming this. 

But since the 404 is also reported in the api server audit logs, doesn't this mean that the 404 was a kubernetes constructed 404 message? 
 
> Does this client make discovery API calls as part of the call returning a NotFound error, and if those discovery calls return 404, would that surface as a 404?

No. This client doesn't make any discovery calls. The client is constructed with the necessary information to be able to request specific custom resources from the api server. 

> You don't see any 404 responses recorded in the apiserver log for those resources (either the individual objects or the discovery APIs for that API group or API group/version)?

The client isn't making any discovery calls. 
And there aren't any specific messages regarding the particular resources for which the audit logs reported a 404 response from the api server.

**Comment by liggitt** [comment]
oh, sorry... I missed the 404s were in the audit logs... do you have an example audit entry for that request?

**Comment by kapad** [comment]
My current suspicion is the following, but my understanding of the kubernetes code base is minimal (at best). 

* apiserver and client are running normally, 
* client is able to request the custom resources, api server responds with 200s, 
* apiserver load increases, response times increase, apiserver restarts, 
* post restart, api server is working to cache the custom resource definitions, 
* client requests custom resource before apiserver has completed caching of CRDs, 
* apiserver responds with 404, 
* within ~30s, client requests the same custom resource again, 
* apiserver responds with 200, 

> per the audit logs, the same external client received, 200, 404, 200, 404, 200, 200, 200, .... (at ~30s intervals).

Regarding this specific sequence, I think the two 404s are from different apiserver instances. And the audit logs that are available do not have any information regarding which instance of the apiserver reported any particular audit message. 

Again, I don't have a deep understanding of this code base. Does the above seem plausible?

**Comment by kapad** [comment]
> oh, sorry... I missed the 404s were in the audit logs... do you have an example audit entry for that request?

Sure. I can share this tomorrow. It's a bit late into the night where I live, and I'll need to clean up the audit logs to remove references that cannot be shared publicly in this issue thread.

**Comment by kapad** [comment]
>     * was it a get or a list request that returned a 404?

request was a GET for a specific resource. 
 
>     * is there information about the timestamps of when the 404 requests started and completed, relative to timestamps of requests for the same resources that got 200 responses?

Yes. This is available. Will share in my AM tomorrow. 

>     * was the request that got a 404 opting into using cached data with a resourceVersion=0 query parameter?

I'm not sure. Can you point me to some docs regarding `opting into using cached data with a resourceVersion=0 query parameter`? I'll need to check the client logic a bit further to confirm this.

**Comment by liggitt** [comment]
https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list has some details... by doing a `get` with a `resourceVersion=0` query parameter, the client explicitly opts into having the result served from the API server's cache, which may be stale. I don't know if controller-runtime clients make get requests with that option enabled or not.

**Comment by kapad** [comment]
> https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list has some details... by doing a `get` with a `resourceVersion=0` query parameter, the client explicitly opts into having the result served from the API server's cache, which may be stale. I don't know if controller-runtime clients make get requests with that option enabled or not.

* The client is setup using the the controller-runtime package using [this constructor](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.14.1/pkg/client/client.go#L66-L78)
* Since the resource is a go type that the client has vendored, the request uses (probably - i'm following through in the code) the typed client. ([GET](https://github.com/kubernetes-sigs/controller-runtime/blob/84c5c9f4e29bfd3cead5c89552c9bc7dc90c45ad/pkg/client/typed_client.go#L137-L150) request)
* When making a request for any particular custom resource, no specific GET options are being set. 
* And looking at [this code](https://github.com/kubernetes/client-go/blob/v0.26.1/rest/request.go#L363-L388), _I don't think_ the request would have any query parameter with `resourceVersion=0`

**Comment by kapad** [comment]
## Audit logs

### Event format and keys
I ran the following command for 1 event to list the keys that are available in the system audit logs. 
```bash
head -n 1 events.txt | jq ".. | keys?" -
```

#### output
```text
[
  "annotations",
  "apiVersion",
  "auditID",
  "cluster_id",
  "kind",
  "level",
  "objectRef",
  "requestReceivedTimestamp",
  "requestURI",
  "responseStatus",
  "sourceIPs",
  "stage",
  "stageTimestamp",
  "user",
  "userAgent",
  "verb"
]
[
  "groups",
  "uid",
  "username"
]
[
  0,
  1,
  2
]
[
  0,
  1
]
[
  "apiGroup",
  "apiVersion",
  "name",
  "resource"
]
[
  "code",
  "metadata"
]
[]
[
  "authentication.k8s.io/legacy-token",
  "authorization.k8s.io/decision",
  "authorization.k8s.io/reason"
]
```

### Kind
The inconsistency was observed for custom resources of the same kind. This resource kind represents cluster level custom resources. (i.e. these custom resources are not namespaced). 

### Resources
In total, the inconsistency regarding 404s was observed for a total of 3 custom resources (of a total of about 10k custom resources). 
The audit events for each resource are attached in a different file, and to prevent leaking private organization information, I've renamed these `resourceA`, `resourceB`, and `resourceC`. 

#### Clients
The client that is requesting `resoourceC`, makes requests at a lower frequency than the clients that request `resourceA` and `resourceB`. 

Additionally, each resource above is requested for by a different client instance. There's a 1:1 mapping from the client instance that's requesting (polling) a resource, and a custom resource. (i.e. each client is only interested in requesting a resource with a specific key, and this is configured when instantiating the client).

### Events
Again, to prevent leaking any private organization information, I've filtered events for each resource using the following jq
```text
{
    kind: .kind, 
    apiVersion: .apiVersion,
    level: .level,
    auditID: .auditID,
    stage: .stage,
    requestURI: "/apis/<customresourcegroup>/v1/<customresourcekind>/<customresourcename>",
    verb: .verb,
    httpResponseCode: .responseStatus.code,
    requestReceivedTimestamp: .requestReceivedTimestamp, 
    stageTimestamp: .stageTimestamp
[resourceC.events.txt](https://github.com/kubernetes/kubernetes/files/10487655/resourceC.events.txt)
[resourceB.events.txt](https://github.com/kubernetes/kubernetes/files/10487656/resourceB.events.txt)
[resourceA.events.txt](https://github.com/kubernetes/kubernetes/files/10487657/resourceA.events.txt)

}
```

**Comment by kapad** [comment]
Thanks for following up on this issue. 

There's some more analysis that's been done during the outage window. I hope this helps with identifying the root cause for this issue. 

### Api server pod restart timestamps
```
cp-0: 2023-01-20T13:55:55.411442194Z
cp-1: 2023-01-20T13:55:57.304382498Z
cp-2: 2023-01-20T13:55:55.142213538Z
cp-3: 2023-01-20T13:56:00.17593755Z
cp-4: 2023-01-20T13:56:04.519953223Z
```

Additionally, the apiserver requests from the external clients use cluster ingress, and using the gateway logs show that all 3 of the inconsistent 404s were from the same apiserver instance (`cp-3`). 
logs: [ingress.txt](https://github.com/kubernetes/kubernetes/files/10499412/ingress.txt)

### inconsistent responses
The `cp-3` api server instance restarted at `2023-01-20T13:56:00.17593755Z`, and the inconsistent 404s had the following timestamps
* `2023-01-20T13:56:00.987Z`
* `2023-01-20T13:56:04.791Z`
* `2023-01-20T13:56:02.328Z`

### Faulty api server logs
I'm attaching the logs of the api server pod which responded with inconsistent 404s, for 30s prior to the restart, and about a minute after the restart. This timeframe also includes the timeframe in which inconsistent 404s were seen. 
logs: [cp-3-logs.txt](https://github.com/kubernetes/kubernetes/files/10502107/cp-3-logs.txt)

#### Note on logs
* `<customresourcekind>` refers to the custom resource for which we saw some inconsistencies.
* `<customresourcekind-B>` to `<customresourcekind-J>` are different custom resource kinds for which we did not see any inconsistency. 
* the api server logs do not have any logs for `resourceA`, `resourceB` or `resourceC` for which inconsistent 404s were returned.
* `<somecustomresourcename>` has been reused multiple times as a placeholder for resources of various different names.

**Comment by liggitt** [comment]
Thanks for the info... a few more questions:

1. are the audit logs from https://github.com/kubernetes/kubernetes/issues/115241#issuecomment-1401550482 and the server logs from https://github.com/kubernetes/kubernetes/issues/115241#issuecomment-1404002194 from the same occurrence?

> In total, the inconsistency regarding 404s was observed for a total of 3 custom resources (of a total of about 10k custom resources).

2. Were the spurious 404s seen for three custom resource _instances_ of the same type, or different types?

3. Do you have other custom resource types in the same API group/version as the one returning 404s? If so, were any other requests made to those other types in the same API group/version against the same server that returned the 404s?

There are two objects the API server needs to fill caches for before it can reliably serve CRD requests:
* APIService objects (controls routing for a particular API group/version) - successful requests to the same group/version (but a different resource type) after server restart but before the 404s were encountered would indicate this is not the issue
* CustomResourceDefinition objects (controls serving of a particular group/version/resource type)

It looks like the logic to return 503s until both of these are synced got reworked in 1.23 in https://github.com/kubernetes/kubernetes/pull/104748 ... cc @p0lyn0mial @sttts on this issue

Additionally, it looks like the logic relies on HasSynced, which I know @lavalamp just reworked in 1.27 in https://github.com/kubernetes/kubernetes/pull/113985 (the CRD handling aspect shouldn't be impacted, since it looks directly at the store... I'm less sure about the APIService aspect, which might have relied on )

**Comment by lavalamp** [comment]
Without reading all the conversation, I think what is going on is a load balancer (I know you have one otherwise there would not be 5xx responses while apiserver was down) is sending CR traffic to an apiserver which hasn't finished loading CRDs.

Does your load balancer perform health checks before forwarding traffic to apiserver? I believe such an apiserver should not be returning true to a /readyz request.

**Comment by kapad** [comment]
>     1. are the audit logs from [Potential inconsistent behaviour with apiserver apiextensions (custom resources) and 404s #115241 (comment)](https://github.com/kubernetes/kubernetes/issues/115241#issuecomment-1401550482) and the server logs from [Potential inconsistent behaviour with apiserver apiextensions (custom resources) and 404s #115241 (comment)](https://github.com/kubernetes/kubernetes/issues/115241#issuecomment-1404002194) from the same occurrence?

Yes. The audit logs and the server logs are from the same occurrence. 
Also, to add some details, 
* Clients connect to the k8s apiserver via an istio ingress gateway.
* From the gateway logs - `ingress.txt` file in https://github.com/kubernetes/kubernetes/issues/115241#issuecomment-1404002194
    * we identified that the inconsistent 404s were all from the same server instance. 
    * The file - `cp-3-logs.txt` in the same comment has the server logs for this single instance. 
    * These server logs represent the same time window during which we saw the inconsistent 404s. 
 
>     2. Were the spurious 404s seen for three custom resource _instances_ of the same type, or different types?

The same type.

>     3. Do you have other custom resource types in the same API group/version as the one returning 404s? If so, were any other requests made to those other types in the same API group/version against the same server that returned the 404s?

Yes. We have other custom resource types in the same API group/version. But we did not see any other requests (other than the 3 mentioned) receiving inconsistent 404s.

**Comment by kapad** [comment]
> Does your load balancer perform health checks before forwarding traffic to apiserver? I believe such an apiserver should not be returning true to a /readyz request.

Yes. We are using istio ingress gateway as a load balancer/cluster ingress solution. I believe the istio ingress gateway 

From my comment earlier: 

> The cp-3 api server instance restarted at `2023-01-20T13:56:00.17593755Z`, and the inconsistent 404s had the following timestamps
> 
> * `2023-01-20T13:56:00.987Z`
> * `2023-01-20T13:56:04.791Z`
> * `2023-01-20T13:56:02.328Z`

Also, etcd was under heavy load during this same time window. The api server logs have several `TRACE` level messages regarding etcd response time slowness.

This following log message, leads me to think that the crd cache was not ready when responding to the above requests. 
```
I0120 13:56:04.903983       1 shared_informer.go:247] Caches are synced for crd-autoregister
```

**Comment by kapad** [comment]
> [EDIT: no, extensions-apiserver returns 503 before it has synced]
So, does this mean that the expected response of the api-server for custom resource requests would be a 503 in case the request is made prior to the cache sync? (Does this hold true even in cases where the api-server is not returning `200` for `/readyz` ?)

**Comment by lavalamp** [comment]
> does this mean that the expected response of the api-server for custom resource requests would be a 503 in case the request is made prior to the cache sync? 

Not really because there are a bunch of things that sync; the thing that matters is the part of extension apiserver that installs handlers for all CR types, not e.g. the watch cache.

**Comment by lavalamp** [comment]
I did look at the logs & code a while back, btw, and I couldn't find any answers before I had to do something else.

I don't expect the problem to be in the apiservice handler since those objects are persisted to etcd, but I guess it's possible, I don't have any other theories either.

Unfortunately we probably will need a way to reproduce to fix this.

**Comment by aaron-prindle** [comment]
/triage accepted
@kapad - are you able to provide repro steps here to allow for further investigation?
